src/sage/combinat/tuple.py

```diff
@@ -75,7 +75,7 @@ def __init__(self, S, k):
         """
         self.S = S
         self.k = k
-        self._index_list = [S.index(s) for s in S]
+        self._index_list = list(dict.fromkeys(S.index(s) for s in S))
```

Comment:
Why not simply using `self._index_list = list(range(len(S)))` ?

Reply:
may be `self._index_list = range(len(S))` is enough

Reply:
We need to remove duplicates from the list (that is the  bug noted in the conversation in the bug report).  The reason that I use the `dict.fromkeys` dance instead of `set` is so that the order of elements is not changed, leaving the order of the generated tuples the same as before.  (The same is true of the seemingly gratuitous `reverse` in the `__iter__`)

The reason for the `_index_list` is to remove duplicates from a list that has non-hashable elements.

range(len(S)) absolutely won't work.

Reply:
I see. Thanks for the precise answer.

Reply:
I think it is very bad to rely on the `dict` order behavior (which could change without notice) and it is not clear that this is what the code is trying to do. IMO it would be much better to then run `sorted(set(self._index_list))` to get the unique elements.

Reply:
We still do the index list dance so we are just hashing integers regardless.  All using `dict` bought us was a guarantee of preserving the order at the cost of some clarity in the code.

Reply:
They are just `int`s. Please read what I actually wrote more carefully.

Reply:
I know the order is currently guaranteed, but that is basically an implementation detail that could change (as it is not the fundamental programming model for a `dict`). It also makes for very brittle code a change the insertion order means everything gets broken.

Reply:
The order is guaranteed precisely so that one can rely on it in code.

Reply:
Even if there is now a strong guarantee that it won't change (which I would not say there is considering Python has shown it is okay breaking backwards incompatibility), it still is brittle and obfuscated code. Having an explicit `sorted` makes the intent of the code clear.

Reply:
> considering Python has shown it is okay breaking backwards incompatibility

If you're referring to the Python 2 -> Python 3 transition, well that won't happen again.